_posts/2019/2019-07-24-how-to-transpile-node-modules-with-babel-and-webpack-in-a-monorepo.md

```diff
@@ -0,0 +1,140 @@
+
+### A story of Javascript standards and transpilation
+
+In this case though, we are not transpiling between Javascript and a derivative language, but between 2 standards of Javascript- ES5 and ES6.
+The ES stands for ECMAScript, which is a [scripting-language specification](https://en.wikipedia.org/wiki/ECMAScript){:target="_blank"}, created to standardize Javascript.
+ES5 refers to the fifth edition of ECMAScript, which is currently supported by all modern browsers.
+ES6- also known as ES2015, is the sixth edition and added several helpful features to Javascript which can be found [here](http://es6-features.org){:target="_blank"}.
+Now, to avoid browser compatibility issues, it is generally advisable to transpile any code meant to run in the browser and written in ES6 or greater, to ES5.
+This is because while most browsers have added support for ES6, there are still some holdouts (I'm looking at you, Internet Explorer!!) with quite widespread usage.
+Other ways of ensuring browser compatibility- which can of course be used with transpilation (and should be in most cases) include avoiding or [polyfilling](https://en.wikipedia.org/wiki/Polyfill_(programming)){:target="_blank"} browser or DOM apis not supported in a browser, or [feature detection](https://en.wikipedia.org/wiki/Feature_detection_(web_development)){:target="_blank"}.
+
+### Tools we use
+
+There are a number of tools for Javascript transpilation out there, but we use [Babel](https://babeljs.io/docs/en/){:target="_blank"} - an extremely powerful Javascript transpilation toolchain that supports conversion between Javascript standards as well as its derivative languages.
+Babel can be configured using a `.babelrc` file, or a `babel.config.js` file.
+Another tool we rely on heavily in our build environment is [Webpack](https://webpack.js.org/concepts){:target="_blank"}, a static module bundler for modern JavaScript applications.
+Webpack is highly configurable, and can be set up so one can configure Babel from within the webpack configuration file, which is usually called `webpack.config.js` using the Babel plugin.
+
+## So what happened then?
+
+We do not transpile the contents of our `node_modules` folder, to reduce the size of our bundle and avoid extremely long build times.
+This is because we expect that we receive already transpiled code from our dependencies.
+This time, we were wrong.
+
+## What we did...
+
+Of course, we cannot transpile our entire `node_modules` folder, since that would increase build time by a terrible factor.
+The compromise we came up with was to modify the regex pattern we used to exclude the `node_modules` folder in our webpack/babel config so that it excluded everything in the `node_modules` folder _except the folder containing the module that was not transpiled_.
+
+Essentially we changed this:
+
+```
+...
+  exclude: /node_modules/,
+...
+```
+
+into this:
+
+```
+...
+  exclude: /node_modules\/?!name-of-untranspiled-module.*/,
```

Comment:
Hmm, take a look:
Shouldn't this regex match the string `node_modules/slpalspalspalsa`?
:confused: 

```javascript
regex = /node_modules\/?!name-of-untranspiled-module.*/
str = 'node_modules/slpalspalspalsa'
str.match(regex)  // null
```

Reply:
The negative lookahead needs wrappings parens

Reply:
Using https://regex101.com/:

Testing the string: `node_modules/slpalspalspalsa` with regex `node_modules\/?!name-of-untranspiled-module.*` -> no matches.
![Screenshot from 2019-08-07 15-29-04](https://user-images.githubusercontent.com/17606011/62626639-1db8ee00-b928-11e9-85fe-f17a2734dd53.png)

Using regex `node_modules\/(?!(name-of-untranspiled-module))`
String `node_modules/slpalspalspalsa` -> 1 match
![Screenshot from 2019-08-07 15-33-02](https://user-images.githubusercontent.com/17606011/62627041-eb5bc080-b928-11e9-97bf-e1a403722512.png)


String `node_modules/name-of-untranspiled-module` -> no matches.

![Screenshot from 2019-08-07 15-33-13](https://user-images.githubusercontent.com/17606011/62627025-e39c1c00-b928-11e9-8d98-a32893b1b152.png)

The expected result, so far.